private/bufpkg/bufpolicy/digest.go

```diff
@@ -212,3 +218,207 @@ func (d *digest) String() string {
 }
 
 func (*digest) isDigest() {}
+
+// getO1Digest returns the O1 digest for the given PolicyConfig.
+func getO1Digest(policyConfig PolicyConfig) (Digest, error) {
+	policyDataJSON, err := marshalStablePolicyConfig(policyConfig)
+	if err != nil {
+		return nil, err
+	}
+	bufcasDigest, err := bufcas.NewDigestForContent(bytes.NewReader(policyDataJSON))
+	if err != nil {
+		return nil, err
+	}
+	return NewDigest(DigestTypeO1, bufcasDigest)
+}
+
+// marshalStablePolicyConfig marshals the given PolicyConfig to a stable JSON representation.
+//
+// This is used for the O1 digest and should not be used for other purposes.
+// It is a valid JSON encoding of the type buf.registry.policy.v1beta1.PolicyConfig.
+func marshalStablePolicyConfig(policyConfig PolicyConfig) ([]byte, error) {
+	lintConfig := policyConfig.LintConfig()
+	if lintConfig == nil {
+		return nil, syserror.Newf("policyConfig.LintConfig() must not be nil")
+	}
+	breakingConfig := policyConfig.BreakingConfig()
+	if breakingConfig == nil {
+		return nil, syserror.Newf("policyConfig.BreakingConfig() must not be nil")
+	}
+	pluginConfigs, err := xslices.MapError(policyConfig.PluginConfigs(), func(pluginConfig bufconfig.PluginConfig) (*policyV1Beta1PolicyConfig_PluginConfig, error) {
+		ref := pluginConfig.Ref()
+		if ref == nil {
+			return nil, fmt.Errorf("PluginConfig must have a non-nil Ref")
+		}
+		optionsConfig, err := optionsToOptionConfig(pluginConfig.Options())
+		if err != nil {
+			return nil, err
+		}
+		return &policyV1Beta1PolicyConfig_PluginConfig{
+			Name: policyV1Beta1PolicyConfig_PluginConfig_Name{
+				Owner:  ref.FullName().Owner(),
+				Plugin: ref.FullName().Name(),
+				Ref:    ref.Ref(),
+			},
+			Options: optionsConfig,
+			Args:    pluginConfig.Args(),
+		}, nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed converting PluginConfigs to PolicyConfig_CheckPluginConfig: %w", err)
+	}
+	slices.SortFunc(pluginConfigs, func(a, b *policyV1Beta1PolicyConfig_PluginConfig) int {
```

Comment:
@unmultimedio added plugin sorting. The other cases:
- Plugin options are sorted [here](https://github.com/bufbuild/buf/pull/3889/files#diff-88c2427302a8a8532fefea9c35818ac1173617eff9215a925948cbf00664388aR361-R363). The map representation does not maintain the original order in the Config.
- Use and Except are defined as sorted, see [here](https://github.com/bufbuild/buf/blob/78837d6a793aa4a28e9817a60d51bfe3b1e01f08/private/bufpkg/bufconfig/check_config.go#L87-L90)